### PR TITLE
GMC: add prepocessUserConfigmap to change the hardcode endpoint in the manifests

### PR DESCRIPTION
**Pull request by KfreeZ**
## Description

As the configmap file in the manifest is hardcode, the endpoint's service name and namespace cannot be changed.
This PR read the ns from context and service name/port from the services manifest files, create the adjusted config map

## Issues

n/a

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies

n/a

## Tests

after the controller reconciliation, the configmap is updated
```
root@cis-gms-worker-3:~/zkf/dev/GenAIInfra/microservices-connector# kubectl get cm -n chatqa 
NAME               DATA   AGE
kube-root-ca.crt   1      3d9h
qna-config   13     9m41s
```

for xeon
```
root@cis-gms-worker-3:~/zkf/dev/GenAIInfra/microservices-connector# kubectl get cm -n chatqa qna-config-xeon -o yaml
apiVersion: v1
data:
  EMBEDDING_MODEL_ID: BAAI/bge-base-en-v1.5
  EMBEDDING_SERVICE_HOST_IP: embedding-svc
  HUGGINGFACEHUB_API_TOKEN: insert-your-huggingface-token-here
  INDEX_NAME: rag-redis
  LLM_MODEL_ID: Intel/neural-chat-7b-v3-3
  LLM_SERVICE_HOST_IP: llm-svc
  **REDIS_URL: redis://redis-vector-db.default.svc.cluster.local:6379**
  RERANK_MODEL_ID: BAAI/bge-reranker-large
  RERANK_SERVICE_HOST_IP: reranking-svc
  RETRIEVER_SERVICE_HOST_IP: retriever-svc
  **TEI_EMBEDDING_ENDPOINT: http://tei-embedding-svc.chatqa.svc.cluster.local:6006
  TEI_RERANKING_ENDPOINT: http://tei-reranking-svc.chatqa.svc.cluster.local:8808
  TGI_LLM_ENDPOINT: http://tgi-svc.chatqa.svc.cluster.local:9009**
kind: ConfigMap
metadata:
  creationTimestamp: "2024-06-06T16:38:46Z"
  name: qna-config-xeon
  namespace: chatqa
  resourceVersion: "1014630"
  uid: 53fefcf9-f4ec-4f6c-9ab4-874fa264c4c5
```

### CODE REVIEW

File: microservices-connector/internal/controller/gmconnector_controller.go [hunk 1]
```diff
@@ -431,6 +458,135 @@ func (r *GMConnectorReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+		var yamlMap map[string]interface{}
+		err = yaml2.Unmarshal(yamlData, &yamlMap)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal YAML data: %v", err)
+		}
+		cfgmapName := yamlMap["metadata"].(map[interface{}]interface{})["name"].(string) + "-" + hwType
+		yamlMap["metadata"].(map[interface{}]interface{})["name"] = cfgmapName
+		if data, ok := yamlMap["data"].(map[interface{}]interface{}); ok {
+			// Update the value of "TEI_EMBEDDING_ENDPOINT" field
+			if _, ok := data["TEI_EMBEDDING_ENDPOINT"].(string); ok {
```
Comment (irisdingbj) [inline microservices-connector/internal/controller/gmconnector_controller.go:498]: please add REDIS_URL handling as well.

File: microservices-connector/internal/controller/gmconnector_controller.go [hunk 2]
```diff
@@ -297,7 +302,7 @@ type TeiEmbeddingCfg struct {
 	HttpsProxy       string
 }
 
-type RetriverCfg struct {
+type RetrieverCfg struct {
```
Comment (irisdingbj) [inline microservices-connector/internal/controller/gmconnector_controller.go:305]: Is it possible to remove all these structs?  We can apply what user defined in config directly:

          config:
            endpoint: /v1/chat/completions
            hostPath: /root/GMC/data/tgi
            modelId: ise-uiuc/Magicoder-S-DS-6.7B

**Comment by zhlsunshine** [comment]
This PR is related to https://github.com/opea-project/GenAIInfra/issues/77

File: microservices-connector/internal/controller/gmconnector_controller.go [hunk 3]
```diff
@@ -431,6 +458,135 @@ func (r *GMConnectorReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+			return fmt.Errorf("failed to unmarshal YAML data: %v", err)
+		}
+		cfgmapName := yamlMap["metadata"].(map[interface{}]interface{})["name"].(string) + "-" + hwType
+		yamlMap["metadata"].(map[interface{}]interface{})["name"] = cfgmapName
+		if data, ok := yamlMap["data"].(map[interface{}]interface{}); ok {
+			// Update the value of "TEI_EMBEDDING_ENDPOINT" field
+			if _, ok := data["TEI_EMBEDDING_ENDPOINT"].(string); ok {
+				svcName, port, err := getServiceDetails(embdManifest)
+				if err == nil {
+					data["TEI_EMBEDDING_ENDPOINT"] = fmt.Sprintf("http://%s.%s.svc.cluster.local:%d", svcName, ns, port)
```
Comment (zhlsunshine) [inline microservices-connector/internal/controller/gmconnector_controller.go:501]: We may need to extract all these environment variables, then handle them together.

File: microservices-connector/internal/controller/gmconnector_controller.go [hunk 4]
```diff
@@ -431,6 +458,135 @@ func (r *GMConnectorReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	return ctrl.Result{}, nil
 }
 
+// get the svc NamespaceName from the svcCfg and replace the ones in the original config map
+func preprocessUserConfigmap(ns string, hwType string) error {
```
Comment (zhlsunshine) [inline microservices-connector/internal/controller/gmconnector_controller.go:462]: As we discussed,  `preprocessUserConfigmap` is a common function to handle the `configmap` from oneclick, so @KfreeZ will draw a new function, such as named `AAA`,  to handle the hardcode endpoint. And I will refactor the `getCustomConfig` to overwrite the `configmap` based on the service config provided by GMC CR yaml.

File: microservices-connector/internal/controller/gmconnector_controller.go (hunk 2 above)
Comment (zhlsunshine) [reply microservices-connector/internal/controller/gmconnector_controller.go:305]: Hi @irisdingbj, `endpoint` should be leaved for http verb between microservices. `hostPath` can be removed, and `modelId` should be provided by oneclick's configmap(They may create a configmap yaml if not existing).

File: microservices-connector/internal/controller/gmconnector_controller.go (hunk 1 above)
Comment (KfreeZ) [reply microservices-connector/internal/controller/gmconnector_controller.go]: done

**Comment by KfreeZ** [comment]
#91 is merged, close this one.
issue https://github.com/opea-project/GenAIInfra/issues/92 is created to track the coverage problem.

**Pull request closed without merging**